File: aws-python-sdk/examples/put-object.py
"""
This script uploads a file to objects-lite using the AWS SDK for Python (Boto3) with put_object.

Usage:
    python put-object.py --endpoint <endpoint_url> --bucket <bucket_name> --key <object_key> --file <file_path>

Arguments:
    --endpoint : The endpoint URL of the objects-lite service.
    --bucket   : The name of the bucket where the object will be uploaded.
    --key      : The key (name) of the object to be created in the bucket.
    --file     : The path to the file to be uploaded.

The script will prompt for a username and password, which will be encoded in Base64 and used as the access key and
secret key.

Example:
    python put-object.py --endpoint https://<pc-ip>:9440/api/prism/v4.0/objects/ --bucket mybucket --key myfile.txt
    --file /path/to/file.txt

Response Details:
    After a successful upload, the script will print the HTTP status code and ETag of the uploaded object.

Limitation:
    - This script only supports uploading objects smaller than 5 GB. For larger files, consider using multipart upload.
"""

import argparse
import sys
from os.path import dirname, abspath
import urllib3

# Add parent directory to sys.path
sys.path.insert(0, dirname(dirname(abspath(__file__))))

# Import functions from utils.py
from utils import get_credentials, create_s3_client, put_object

# Disable SSL verification warning
urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

def main():
    # Define command-line arguments
    parser = argparse.ArgumentParser(description='Upload a file to objectslite')
    parser.add_argument('--endpoint', required=True, help='Objectslite endpoint')
    parser.add_argument('--bucket', required=True, help='Bucket name')
    parser.add_argument('--key', required=True, help='Key name')
    parser.add_argument('--file', required=True, help='File to upload')
    args = parser.parse_args()

    # Get encoded credentials
    encoded_credentials = get_credentials()

    # Create S3 client
    s3_client = create_s3_client(args.endpoint, encoded_credentials)

    # Perform put_object operation
    put_object(s3_client, args.file, args.bucket, args.key)

if __name__ == "__main__":
    main()

File: aws-java-sdk/src/main/java/com/examples/PutObject.java
/**
 * PutObject.java
 *
 * This class provides functionality to upload a file to Object Lite using the AWS SDK for Java via PutObject API.
 *
 * Usage:
 * mvn clean package
 * java -cp target/aws-java-sdk-examples-1.0.jar com.examples.PutObject --endpoint <endpoint> --bucket <bucket> --key \
 * <key> -file <file>
 *
 * Command Line Options:
 * -e, --endpoint      The endpoint URL of the objects-lite service.
 * -b, --bucket        The name of the bucket where the file will be uploaded.
 * -k, --key           The key (name) of the object to be created in the bucket.
 * -f, --file          The path to the file to be uploaded.
 *
 * Example:
 * java -cp target/aws-java-sdk-examples-1.0.jar com.examples.PutObject --endpoint \
 * https://<pc-ip>:9440/api/prism/v4.0/objects/ --bucket mybucket --key myfile.txt -file /path/to/myfile.txt
 *
 * Response:
 * After the file is uploaded successfully, the script prints the ETag of the uploaded object.
 *
 * Limitations:
 * - This script only supports uploading objects smaller than 5 GB. For larger files, consider using multipart upload.
 */

package com.examples;

import com.utils.Utils;
import com.amazonaws.services.s3.AmazonS3;
import org.apache.commons.cli.*;

public class PutObject {
    private AmazonS3 s3Client;

    /**
     * Constructor to initialize the AmazonS3 client.
     *
     * @param encodedCredentials Base64 encoded credentials.
     * @param endpoint The S3 endpoint to connect to.
     * @throws Exception If an error occurs during client creation.
     */
    public PutObject(String encodedCredentials, String endpoint) throws Exception {
        // Initialize the s3Client in the constructor
        this.s3Client = Utils.CreateS3Client(encodedCredentials, endpoint);
    }

    /**
     * Uploads a file to the specified S3 bucket with the given key.
     *
     * @param bucket The name of the S3 bucket.
     * @param key The key under which to store the new object.
     * @param filePath The path to the file to upload.
     * @throws Exception If an error occurs during the upload.
     */
    public void uploadObject(String bucket, String key, String filePath) throws Exception {
        Utils.PutObject(this.s3Client, bucket, key, filePath);
    }

    public static void main(String[] args) {
        // Disable AWS SDK v1 deprecation announcement
        System.setProperty("aws.java.v1.disableDeprecationAnnouncement", "true");

        // Define command line options
        Options options = new Options();

        // Define endpoint option
        Option endpointOption = new Option("e", "endpoint", true, "Endpoint URL");
        endpointOption.setRequired(true);
        options.addOption(endpointOption);

        // Define bucket option
        Option bucketOption = new Option("b", "bucket", true, "Bucket name");
        bucketOption.setRequired(true);
        options.addOption(bucketOption);

        // Define key option
        Option keyOption = new Option("k", "key", true, "Object key");
        keyOption.setRequired(true);
        options.addOption(keyOption);

        // Define file option
        Option fileOption = new Option("f", "file", true, "File to upload");
        fileOption.setRequired(true);
        options.addOption(fileOption);

        // Create command line parser and formatter
        CommandLineParser parser = new DefaultParser();
        HelpFormatter formatter = new HelpFormatter();

        String endpoint = null;
        String bucket = null;
        String key = null;
        String file = null;
        try {
            // Parse command line arguments
            CommandLine cmd = parser.parse(options, args);

            // Get values from command line arguments
            endpoint = cmd.getOptionValue("endpoint");
            bucket = cmd.getOptionValue("bucket");
            key = cmd.getOptionValue("key");
            file = cmd.getOptionValue("file");
        } catch (ParseException e) {
            System.err.println("Failed to parse command line arguments: " + e.getMessage());
            formatter.printHelp("ObjectsLitePutObject", options);
            System.exit(1);
        }

        // Get credentials from the user
        String encodedCredentials = Utils.GetCredentials();

        try {
            // Create PutObject instance and upload the file
            PutObject putObject = new PutObject(encodedCredentials, endpoint);
            putObject.uploadObject(bucket, key, file);
        } catch (Exception e) {
            System.err.println("Failed to upload object: " + e.getMessage());
            System.exit(1);
        }
    }
}
